### PR TITLE
[memtracker] improve throw bad_alloc logic

### PR DESCRIPTION
**Pull request by codesigner**
## What type of PR is this?
- [ ] bug
- [ ] feature
- [x] enhancement

#### Issue(s) number: 
#5081 

#### Description:

- the overloaded new/delete operator will influence all thread, include third-party's, like rocksdb's flush/compact threads, thrift rpc threads, which is not what we want.
- introduce a throwOnMemoryExceeded in ThreadMemoryStats, it indicat whether throw on memory exceed, default to FALSE, then all thread is default will not get bad_alloc, unless a MemoryCheckGuard is create by that thread;
- MemoryCheckGuard:  used when some routine need mem check turn on, when this guard was added to our code path, will turn on that code scope's memory check, until it freed (turn off on destruction);

## Checklist:
Tests:
- [ ] Unit test(positive and negative cases)
- [ ] Function test
- [ ] Performance test
- [ ] N/A

Affects:
- [ ] Documentation affected (Please add the label if documentation needs to be modified.)
- [ ] Incompatibility (If it breaks the compatibility, please describe it and add the label.）
- [ ] If it's needed to cherry-pick (If cherry-pick to some branches is required, please label the destination version(s).)
- [ ] Performance impacted: Consumes more CPU/Memory

## Release notes:

Please confirm whether to be reflected in release notes and how to describe:
> ex. Fixed the bug .....

### CODE REVIEW

File: src/graph/executor/algo/BFSShortestPathExecutor.cpp [hunk 1]
```diff
@@ -29,18 +29,26 @@ folly::Future<Status> BFSShortestPathExecutor::execute() {
...
+    memory::MemoryCheckGuard guard;
+    return buildPath(true);
+  });
   futures.emplace_back(std::move(leftFuture));
   futures.emplace_back(std::move(rightFuture));
 
   return folly::collect(futures)
       .via(runner())
       .thenValue([this](auto&& status) {
+        memory::MemoryCheckGuard guard;
```
Comment (Shylock-Hg) [inline src/graph/executor/algo/BFSShortestPathExecutor.cpp:46]: I think you could enable it in `Schedueler::execute` and disable it in callback of `Future`.

Comment (codesigner) [reply src/graph/executor/algo/BFSShortestPathExecutor.cpp:46]: > I think you could enable it in `Schedueler::execute` and disable it in callback of `Future`.

`QueryInstance::execute` is the caller of `Schedueler::execute`, it already have MemoryCheckGuard on, see change in QueryInstance.cpp

The thread invoke thenValue's callback was another thread from folly, it should turn MemoryCheckGuard on by itself.

**Comment by critical27** [comment]
Maybe need to consider adding`MemoryCheckGuard` to the `thenValue` in `StorageClientBase::collectResponse/getResponse`. That is where we may encounter oom when retrieving data from storage at the first place.

**Comment by codesigner** [comment]
> Maybe need to consider adding `MemoryCheckGuard` to the `thenValue` in `StorageClientBase::collectResponse/getResponse`. That is where we may encounter oom when retrieving data from storage at the first place.

OK, let me check.

**Review by czpmango — DISMISSED** [review]

**Comment by codesigner** [comment]
After this, I will mannually add some throw bad_alloc logic, to verify code path we turned on should successfully failed that query.

> Maybe need to consider adding `MemoryCheckGuard` to the `thenValue` in `StorageClientBase::collectResponse/getResponse`. That is where we may encounter oom when retrieving data from storage at the first place.

added

**Review by czpmango — APPROVED** [review]

**Comment by codesigner** [comment]
> Could this mechanism cover encode/decode process of thrift?

Not quite sure, is the decode/dencode you mean is code in like DataOps-inl.h, if the decode/encode is write by ourselfs, then yes.

**Pull request merged**